gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/MRCompactorJobRunner.java

```diff
@@ -89,28 +92,57 @@ public Void call() throws IOException, ClassNotFoundException, InterruptedExcept
     DateTime jobStartTime = new DateTime(DateTimeZone.forID(this.jobProps.getProp(
         ConfigurationKeys.COMPACTION_TIMEZONE, ConfigurationKeys.DEFAULT_COMPACTION_TIMEZONE)));
+
     if (this.jobProps.getPropAsBoolean(ConfigurationKeys.COMPACTION_JOB_LATE_DATA_MOVEMENT_TASK, false)) {
-      this.copyLateDataFiles(new Path(this.outputPath, "late"), conf);
+      List<Path> lateFilePaths = Lists.newArrayList();
+      for (String filePathString : this.jobProps.getPropAsList(ConfigurationKeys.COMPACTION_JOB_LATE_DATA_FILES)) {
+        if (FilenameUtils.isExtension(filePathString, getApplicableFileExtensions())) {
+          lateFilePaths.add(new Path(filePathString));
+        }
+      }
+      this.copyDataFiles(
+          new Path(this.outputPath, ConfigurationKeys.COMPACTION_LATE_FILES_DIRECTORY), lateFilePaths, conf);
     } else {
       if (this.fs.exists(this.outputPath) && !canOverwriteOutputDir()) {
         LOG.warn(String.format("Output path %s exists. Will not compact %s.", this.outputPath, this.inputPath));
         return null;
       }
-      addJars(conf);
-      Job job = Job.getInstance(conf);
-      this.configureJob(job);
-      this.submit(job);
+      if (this.jobProps.getPropAsBoolean(ConfigurationKeys.COMPACTION_DEDUPLICATION,
+          ConfigurationKeys.DEFAULT_COMPACTION_DEDUPLICATION)) {
+        addJars(conf);
+        Job job = Job.getInstance(conf);
+        this.configureJob(job);
+        this.submit(job);
+      } else {
+        this.fs.mkdirs(this.tmpPath);
+        List<Path> filePaths = Lists.newArrayList();
+        for (FileStatus status : HadoopUtils.listStatusRecursive(this.fs, this.inputPath)) {
+          if (FilenameUtils.isExtension(status.getPath().getName(), getApplicableFileExtensions())) {
+            filePaths.add(status.getPath());
+          }
+        }
+        this.copyDataFiles(this.tmpPath, filePaths, conf);
+      }
       this.moveTmpPathToOutputPath();
     }
     this.markOutputDirAsCompleted(jobStartTime);
     return null;
   }
 
-  private void copyLateDataFiles(Path outputDirectory, Configuration conf) throws IOException {
-    for (String filePath : this.jobProps.getPropAsList(ConfigurationKeys.COMPACTION_JOB_LATE_DATA_FILES)) {
-      Path outPath = new Path(outputDirectory,
-          StringUtils.removeStart(StringUtils.removeStart(filePath, this.inputPath.toString()), "/"));
-      if (!FileUtil.copy(this.fs, new Path(filePath), this.fs, outPath, false, conf)) {
+  private void copyDataFiles(Path outputDirectory, List<Path> inputFilePaths, Configuration conf)
+      throws IOException {
+    for (Path filePath : inputFilePaths) {
+      String fileName = filePath.getName();
+      Path outPath;
+      int fileSuffix = 0;
+      do {
+        outPath = new Path(outputDirectory, String.format("%s.%d.%s",
+            FilenameUtils.getBaseName(fileName), fileSuffix++, FilenameUtils.getExtension(fileName)));
+      } while (this.fs.exists(outPath));
```

Comment:
Since each file has a unique fileSuffix, each file name should be unique. Do you really need this do-while?